proposals/devcontainer-features.md

```diff
@@ -0,0 +1,143 @@
+
+In addition to the list of features included, the file includes the following properties.
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| name | string | Name of the collection |
+| reference | string | Reference information of the repository and path where the code is stored. |
+| version | string | Version of the code. |
+
+In most cases the `feature-collection.json` file can be generated automatically at the moment of creating a release.
+
+## devcontainer.json properties
+
+Features are referenced in `devcontainer.json`  where the `features` tag consists of an array with the ordered list of features to be included in the container image.
+
+The properties are:
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| id | string | reference to the particular feature to be included. |
+| options | object | Type of the option |
+
+The `id` is the main reference point for how to find and download a particular feature. `id` can be defined in any of the following ways:
+
+| Type | Description |
+| :--- | :--- |
+| feature  | A simple name references a feature included with the application used to execute `devcontainer.json` |
+| organization/repository/{feature or collectionl/feature}/@version | Reference to a particular release in a repository |
+| https://<../URI/..>/devcontainer-features.tgz#{optional feature} | Direct reference to a file download. |
+| ./{local-path}  -or-  ../{local-path} | A path relative to devcontainer.json where a feature or feature collection can be found. |
+
+## Authoring
+
+Features can be authored in a number of languages, the most straight forward being bash scripts. If a feature is authored in a different language information about it should be included in the metadata so that users can make an informed choice about it.
+
+Reference information about the application required to execute the feature should be included in `feature.json` in the metadata section.
+
+Applications should default to `/bin/sh` for features that do not include this information.
+
+If the feature is included in a folder as part of the repository that contains `devcontainer.json` no other steps are necessary.
+
+## Release
+
+A release is created when the objective is to have other users use a feature.
+
+A release consists of the following:
+
+1.- Tar file with all the included files for the feature or feature collection.
+2.- A copy of the `feature.json` or `feature-collection.json` file that defines the contents of the tar file with additional information added to validate it:
+
+| Property | Type | Description |
+| :--- | :--- | :--- |
+| version | string | SemVer version of the release. |
+| checksum | string | Checksum of the tar file. |
+
+## Execution
+
+The application that implements features should:
+
+- Features are executed in the order defined in devcontainer.json
+- It should be possible to execute a feature multiple times with different parameters.
```

Comment:
- Should a feature opt-in to supporting multiple runs? Otherwise this will make writing features harder.
- Is it up to the feature to decide which run will be on the PATH by default?
- How would a feature provide a version switcher to change which install is on the PATH?

Reply:
- I agree that features can opt-in to supporting multiple runs, perhaps as a propertie of `devcontianer-feature.json` that we could validate in some way.
- While making changes to the implementation of the current features I added parameters to control it and with true. That reinforces the idea that features need to declare if they support multiple runs or not.
- I feel like that is more a feature of a tool that a user would use at runtime.

Reply:
It would be easier to let a feature add options for installing multiple versions (maybe with some additional support for doing that). That would allow the feature to control installation of all requested versions in a single run instead of having it run multiple times. With that approach IntelliSense in the devcontainer.json would likely include a hint if a feature supports multiple versions (e.g., an `"additionalVersions"` property).

Reply:
Yeah I don't want to block features from receiving multiple values and using them to loop, the idea is that we should also not disallow multiple runs.